activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb

```diff
@@ -759,29 +759,40 @@ def configure_connection
 
           super
 
+          pragmas = @config.fetch(:pragmas, {}).stringify_keys
+
           # Enforce foreign key constraints
           # https://www.sqlite.org/pragma.html#pragma_foreign_keys
           # https://www.sqlite.org/foreignkeys.html
-          raw_execute("PRAGMA foreign_keys = ON", "SCHEMA")
+          @raw_connection.foreign_keys = pragmas.fetch("foreign_keys", true)
           unless @memory_database
             # Journal mode WAL allows for greater concurrency (many readers + one writer)
             # https://www.sqlite.org/pragma.html#pragma_journal_mode
-            raw_execute("PRAGMA journal_mode = WAL", "SCHEMA")
+            @raw_connection.journal_mode = pragmas.fetch("journal_mode", :wal)
```

Comment:
This is one I feel strongly on, that we shouldn't force WAL mode by default.   Perhaps a better solution if Rails wants to be opinionated that your journal mode should be WAL, we should only set this if the database doesn't exist (if your configuring a connection for a sqlite database that doesnt exist the file is created).

This will still throw an error on a readonly database unless the developer explicitly configures the correct wal mode in their database configuration.   It won't be clear to the developer that this is what they need to do from that error message (which is something like "Can not write to a read only database").  It's a bad developer experience.

Embedded databases have immense value, consider the scenario where one embeds canonical data with each build of their application.   For example one may embed a zip code database, and like other dependencies (i.e. Gemfile) these databases are versioned and meant to be read only.   Now in order to support this use case, I as a developer need to undo a whole bunch of default database configuration.

Instead we should default to the journal_mode that's persisted with the database.

Reply:
> Instead we should default to the journal_mode that's persisted with the database.
In what way is a `journal_mode` persisted with the database? And how would one read the `journal_mode` that is persisted with the database from the database?

Reply:
That being said, I do, respectfully, disagree that WAL shouldn't be the default. You are absolutely right that developers should be able to opt-out of the default, but I think it is very much the majority and default use-case to use a read/write SQLite database, and WAL is the strongly recommended journal_mode for a read/write SQLite database. The reality is that effectively no developers will investigate the defaults, or read documentation suggesting to change the generated files. We shouldn't degrade the experience of the vast majority of developers for an edge-case, and I do believe that read-only databases are an edge-case usage (an important usage, yes, but still edge). 

That being said, if we could provide a clearer and better error message with explicit guidance on how to resolve trying to connect to a read-only SQLite database, I believe we absolutely should. Perhaps a good next step is capturing this exact error state (trying to connect to a read-only database when the `journal_mode` hasn't been explicitly changed from WAL) and providing a custom Rails exception instead which guides the developer on the exact problem and the recommended resolution (changing the `journal_mode` explicitly).

Reply:
> In what way is a journal_mode persisted with the database?

From [the docs](https://www.sqlite.org/pragma.html#pragma_journal_mode)

>The WAL journaling mode is persistent; after being set it stays in effect across multiple database connections and after closing and reopening the database.

Reply:
> but I think it is very much the majority and default use-case to use a read/write SQLite database

I would challenge that.  SQLite's main use case is embeddeding databases.  If you're embedding a SQLite database in your web app, you most likely have already configured it correctly or don't have to deal with concurancy so a normal journal mode is fine.

> The reality is that effectively no developers will investigate the defaults

Good opportunity for those bad developers to learn when they start dealing with production issues :)

It's okay to disagree!   Either way being able to configure the pragmas is a huge step forward.   With WAL forced as a default anyone upgrading to Rails 7.1 will have to account for this (or be left wondering why a WAL file is showing up).

In my use cases, I will just keep using a monkey patch to avoid these defaults rather than adding additional configuration.

Reply:
> And how would one read the `journal_mode` that is persisted with the database from the database?

There is [an accessor](https://github.com/sparklemotion/sqlite3-ruby/blob/740ea0992d4dde0d3b7f8efdd2cf118f1edd8269/lib/sqlite3/pragmas.rb#L296) in the SQLite3 gem.   SQLite allows you to read all pragma values.

Reply:
My point is that the `journal_mode` isn't embedded into the database file. It lives at the connection level:
> This pragma queries or sets the journal mode for databases associated with the current [database connection](https://www.sqlite.org/c3ref/sqlite3.html).
> — [SQLite docs](https://www.sqlite.org/pragma.html#pragma_journal_mode)

I just manually tested this, and you can confirm in an IRB session:
```ruby
require 'sqlite3'

db_name = "example.db"

conn_1 = SQLite3::Database.new db_name
conn_1.journal_mode = :truncate
conn_1.close

conn_2 = SQLite3::Database.new db_name
conn_2.journal_mode
# => "delete"
```

Only the WAL `journal_mode` is persisted in the database itself.

Reply:
Correct.   There are also other persistent pragmas settings too.

Reply:
Fair discussion, but I concur that WAL should be the default. We intend for sqlite to be a full option for use as the main application database, and this is a better option for this.

Reply:
Not a hill I care to die on, and I don't disagree - **WAL should be the default.**   

My contention is that we shouldn't force WAL on _existing databases_ which will result in errors if that database is readonly (it's not backwards compatible with Rails <7.1).

The [approach I took](https://github.com/rails/rails/pull/49691/files#diff-a8f97d11d4db16a3bbab7c50b906674e0f932dc32056cf7fb5d415cd7a39016bR11) to make WAL the default in the Rails generator, rather than in the connector.    

That way anyone creating a new Rails app would get WAL by default (and the other sensible settings), but we wouldn't break existing databases that are readonly.

Also worth noting if we did go back to 7.0 behavior and removed the defaulting of WAL in the connector, that would not break any databases that migrated to WAL when 7.1 was released.  Those databases would continue to use WAL since that setting is persistent.